posts/navarroaxel/19.git-rebase-exec.md

```diff
@@ -0,0 +1,62 @@
+---
+title: Git rebase exec to validate every commit
+published: true
+description: How to run unit tests and build tools for every commit when you rebase
+tags: 'git, bash, productivity, tooling'
+cover_image: ./assets/git-rebase-exec.png
+id: 1949938
+---
+
+In the previous post we checked out how the `git rebase` works and how to use the `--interactive` mode to keep your topic branch up-to-date with the target (e.g. `main`). Adding the `exec` command to run the tests and lint tools to check every rebased commit.
+
+Now, let's check how to use`exec` in a short way.
+
+## Exec, exec, exec
+
+If you need to run the build, tests, or a format tool when you are rebasing because you must guarantee that every commit is a valid commit, you can speed it up using the `git rebase --exec` command.
+
+```bash
+git rebase main --exec "npm test" --exec "npm run build"
+```
+
+If the command fails 😩 you need to fix your code and then resume the rebase with `git rebase --continue`.
+
+```bash
+git add .
+git commit --amend
+git rebase --continue
+```
+
+### Reschedule failed exec
+
+If an `exec` command exits with a failure, your rebase will return the control of the shell to you. The `git rebase --continue` will resume the execution but it won't run the failed `exec` again. To avoid this you can use [`--reschedule-failed-exec`](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---reschedule-failed-exec).
+
+```bash
+git rebase main --exec "npm test" --exec "npm run build" --reschedule-failed-exec
+```
+
+You can make this the default behavior using the [rebase.rescheduleFailedExec](https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt-rebaserescheduleFailedExec) setting:
+
+```bash
+git config --global rebase.rescheduleFailedExec true
+```
+
+To turn this off when the `rescheduleFailedExec` is globally `true` you should use the `--no-reschedule-failed-exe` flag:
+
+```bash
+git rebase main --exec "npm test" --no-reschedule-failed-exec
+```
+
+Or, you can unset this globally:
+
+```bash
+git rebase --global --unset rebase.rescheduleFailedExec
+```
+
+## Conclusion
+
+When your commits will live on your Git repository forever (e.g. inside the `main` branch) without being squashed (combined into one) it's useful to guarantee the validity of each commit individually. Git provides this `exec` mechanism to automate this process.
+
+Do you often check the tests and build for every commit of your created pull request? 😉
```

Comment:
Is this the correct question? Sounds a bit weird.
This seems better; keep in mind it's a different one though.
```suggestion
How often do you check the tests and build for every commit of your created pull request? 😉
```